eav/queryset.py

```diff
@@ -1,47 +1,273 @@
+                    # Create model queryset.
+                    _q = model_cls.objects.filter(**{fname: attrval[1]})
+                else:
+                    # Second val in tuple is a queryset.
+                    _q = attrval[1]
+
+                # Explicitly check for None. 'or' doesn't work here
+                # as empty QuerySet, which is valid, is falsy.
+                q = q if q != None else _q
+
+                if expr.connector == 'AND':
+                    q &= _q
+                else:
+                    q |= _q
+
+            # If any two children were merged,
+            # update parent expression.
+            if q != None:
+                expr.children = other + [('pk__in', q)]
+
+    return expr
+
+
+def eav_filter(func):
+    '''
+    Decorator used to wrap filter and exclude methods. Passes args through
+    expand_q_filters and kwargs through expand_eav_filter. Returns the
+    called function (filter or exclude).
+    '''
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        nargs = []
+        nkwargs = {}
+
+        for arg in args:
+            if isinstance(arg, Q):
+                # Modify Q objects (warning: recursion ahead).
+                arg = expand_q_filters(arg, self.model)
+                # Rewrite Q-expression to safeform.
+                arg = rewrite_q_expr(self.model, arg)
+            nargs.append(arg)
+
+        for key, value in kwargs.items():
+            # Modify kwargs (warning: recursion ahead).
+            nkey, nval = expand_eav_filter(self.model, key, value)
+
+            if nkey in nkwargs:
+                # Apply AND to both querysets.
+                nkwargs[nkey] = (nkwargs[nkey] & nval).distinct()
+            else:
+                nkwargs.update({nkey: nval})
+
+        return func(self, *nargs, **nkwargs)
+
+    return wrapper
+
+
+def expand_q_filters(q, root_cls):
+    '''
+    Takes a Q object and a model class.
+    Recursivley passes each filter / value in the Q object tree leaf nodes
+    through expand_eav_filter
+    '''
+    new_children = []
+
+    for qi in q.children:
+        if type(qi) is tuple:
+            # This child is a leaf node: in Q this is a 2-tuple of:
+            # (filter parameter, value).
+            key, value = expand_eav_filter(root_cls, *qi)
+            new_children.append(Q(**{key: value}))
+        else:
+            # This child is another Q node: recursify!
+            new_children.append(expand_q_filters(qi, root_cls))
+
+    q.children = new_children
+    return q
+
+
+def expand_eav_filter(model_cls, key, value):
+    '''
+    Accepts a model class and a key, value.
+    Recurisively replaces any eav filter with a subquery.
+
+    For example::
+
+        key = 'eav__height'
+        value = 5
+
+    Would return::
+
+        key = 'eav_values__in'
+        value = Values.objects.filter(value_int=5, attribute__slug='height')
+    '''
```

Comment:
@IwoHerka style - no newline between end of docstring and first class field. Consistency?